sw/device/silicon_creator/mask_rom/bootstrap.c

```diff
@@ -0,0 +1,321 @@
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/drivers/spi_device.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "flash_ctrl_regs.h"
+#include "gpio_regs.h"
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+#include "otp_ctrl_regs.h"
+
+enum {
+  /*
+   * Maximum flash address, exclusive.
+   */
+  kMaxAddress = FLASH_CTRL_PARAM_BYTES_PER_BANK * 2,
+};
+
+/**
+ * Bootstrap states.
+ *
+ * OpenTitan bootstap consists of three states between which the chip
+ * transitions sequentially.
+ *
+ * Encoding generated with
+ * $ ./util/design/sparse-fsm-encode.py -d 5 -m 3 -n 32 \
+ *     -s 375382971 --language=c
+ *
+ * Minimum Hamming distance: 17
+ * Maximum Hamming distance: 19
+ * Minimum Hamming weight: 16
+ * Maximum Hamming weight: 19
+ */
+typedef enum bootstrap_state {
+  /**
+   * Initial bootstrap state where the chip waits for a SECTOR_ERASE or
+   * CHIP_ERASE command.
+   */
+  kBootstrapStateErase = 0xd4576543,
+  /**
+   * Second bootstrap state where the chip verifies that all data banks have
+   * been erased.
+   */
+  kBootstrapStateEraseVerify = 0xf3c71bac,
+  /**
+   * Final bootstrap state. This is the main program loop where the chip handles
+   * erase, program, and reset commands.
+   */
+  kBootstrapStateProgram = 0xbdd8ca60,
+} bootstrap_state_t;
+
+/**
+ * Handles access permissions and erases both data banks of the embedded flash.
+ *
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_chip_erase(void) {
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolFalse,
+      .erase = kHardenedBoolTrue,
+  });
+  rom_error_t err_0 = flash_ctrl_data_erase(0, kFlashCtrlEraseTypeBank);
+  rom_error_t err_1 = flash_ctrl_data_erase(FLASH_CTRL_PARAM_BYTES_PER_BANK,
+                                            kFlashCtrlEraseTypeBank);
+  flash_ctrl_data_default_perms_set((flash_ctrl_perms_t){
+      .read = kHardenedBoolFalse,
+      .write = kHardenedBoolFalse,
+      .erase = kHardenedBoolFalse,
+  });
+
+  HARDENED_RETURN_IF_ERROR(err_0);
+  return err_1;
+}
+
+/**
+ * Handles access permissions and erases a 4 KiB region in the data partition of
+ * the embedded flash.
+ *
+ * Since OpenTitan's flash page size is 2 KiB, this function erases two
+ * consecutive pages starting with the one at `addr`.
+ *
+ * @param addr Address of the first page to erase, must be page aligned.
+ * @return Result of the operation.
+ */
+static rom_error_t bootstrap_sector_erase(uint32_t addr) {
+  static_assert(__builtin_popcount(FLASH_CTRL_PARAM_BYTES_PER_PAGE) == 1,
+                "Bytes per page must be a power of two.");
+  enum {
+    /**
+     * Mask for checking that `addr` is page aligned.
+     */
+    kPageAddrMask = FLASH_CTRL_PARAM_BYTES_PER_PAGE - 1,
+  };
+
+  if (addr & kPageAddrMask || addr >= kMaxAddress) {
+    return kErrorBootstrapEraseAddress;
+  }
```

Comment:
Is this checking 2KiB or 4KiB alignment?  Since we're emulating 4KiB sectors, this should be checking for 4KiB alignment.

We should check the behavior of a real EEPROM here - I believe the low bits of the address are simply discarded and the page is erased.

Reply:
I'll check the datasheets but PLMK if you find the answer to this or the following comment before me.

Reply:
I checked this behavior on an MX25L512 part.  That EEPROM simply discards the low bits and erases the 4KiB page referred to by the relvant bits of the address.

Reply:
Thanks @cfrantz and @a-will. Updated this part so that `addr` just needs to fall within the 4 KiB region being deleted (truncated to the lower 4 KiB aligned address). I kept the `addr >= kMaxAddress` check though.